text/0000-futures.md

```diff
@@ -154,31 +164,323 @@ can be dropped.
+in other ways.
+
+The relation between those `Waker` types is outlined in the following definitions:
+
+```rust
+/// A `RawWaker` allows the implementor of a task executor to customize the
+/// behavior of `LocalWaker`s and `Waker`s.
+/// It consists of a data pointer and a virtual function pointer table (vtable) that
+/// customizes the behavior of the `RawWaker`.
+#[derive(PartialEq)]
+pub struct RawWaker {
+    /// A data pointer, which can be used to store arbitrary data as required
+    /// by the executor. This could be e.g. a type-erased pointer to an `Arc`
+    /// that is associated with the task.
+    /// The value of this field gets passed to all functions that are part of
+    /// the vtable as first parameter.
+    pub data: *const (),
+    /// Virtual function pointer table that customizes the behavior of this waker.
+    pub vtable: &'static RawWakerVTable,
+}
+
+/// A virtual function pointer table (vtable) that allows to customize the
+/// behavior of a `RawWaker`.
+/// The pointers which get passed to all functions inside the vtable are the
+/// values of the `data` field.
+#[derive(PartialEq, Copy, Clone)]
+pub struct RawWakerVTable {
+    /// This function will be called when the `RawWaker` gets cloned, e.g. when
+    /// the `Waker` or `LocalWaker` in which the `RawWaker` is stored gets cloned.
+    ///
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a `RawWaker` and associated
+    /// task. The implementation must return a valid `RawWaker` that behaves
+    /// equivalent to the `RawWaker` that got cloned. E.g. cloning a `RawWaker`
+    /// that implemented a thread-safe wakeup for use in `Waker` must return
+    /// a `RawWaker` that implements the same wakeup behavior.
+    pub clone: unsafe fn(*const ()) -> RawWaker,
+    /// This function will be called when a `LocalWaker` should be converted into
+    /// a thread-safe `Waker`. The implementation of this function must return
+    /// a new `RawWaker` which can fulfill the requirements of `Waker`. It can
+    /// exchange the vtable for that purpose. E.g. it might replace the `wake`
+    /// function with a varient that supports cross-thread wakeups.
+    ///
+    /// The old `LocalWaker` which contained the data pointer should be consumed
+    /// while performing the operation. After the operation the `LocalWaker`
+    /// won't exist anymore, only the new `Waker`.
+    /// This means that if both instances would utilize the same reference-counted
+    /// data, changing the reference count would not be necessary.
+    ///
+    /// If a conversion is not supported, the implementation should return
+    /// `None`. In this case the implementation must still make
+    /// sure that the data pointer which is passed to the function is correctly
+    /// released, e.g. by calling the associated `drop_fn` on it.
+    pub into_waker: unsafe fn(*const ()) -> Option<RawWaker>,
```

Comment:
Should this function own the data (`*const ()`) and return `Option<Waker>` instead? This way if there are two different waking mechanisms for local and sync wakers it would be supported. Furthmore returning a `Waker`, rather then `RawWaker`, makes it clear that all the requirements of the `Waker` type must be met.

Reply:
@Thomasdezeeuw if there are different implementations then the returned `RawWaker` can contain different function pointers.

Reply:
@Nemo157 Right scratch the first part, I had in my head that `data` was part of `Waker`, not `RawWaker`. What are you thoughts about the second part?

Reply:
I could go either way.  I like the fact that when writing the `RawWaker` type currently you don’t need to import any of the concrete types it will be wrapped in, and think that the documentation is clear enough (and the entire area unsafe enough) that implementors will be careful about what they’re doing. But having to call `Waker::new_unchecked` doesn’t seem like a massive burden (although, my current implementation of `RawWaker` is 100% safe code, this would require adding an `unsafe` block to it).

Reply:
I see it the same as @Nemo157: I like the separation of Api concerns. 
There should also be no extra unsafe call involved, since the new_unchecked from the result would happen in LocalWaker::into_waker()